### PR TITLE
Warn when dashboard is available

### PR DESCRIPTION
**Pull request by timgl**
## Changes

It's confusing when you get sent a link to a dashboard but you've selected a different project, as you'll get a "dashboard not found" message. This PR allows you to easily switch to the correct project.

image

## How did you test this code?

CI + manual QA:
- Have 2 projects
- Copy the link of a dashboard in project 1 and switch to project 2
- Load the URL of project one, see the warning appear

### CODE REVIEW

File: posthog/api/dashboard.py [hunk 1]
```diff
@@ -187,7 +189,18 @@ def get_queryset(self) -> QuerySet:
     def retrieve(self, request: Request, *args: Any, **kwargs: Any) -> response.Response:
         pk = kwargs["pk"]
         queryset = self.get_queryset()
-        dashboard = get_object_or_404(queryset, pk=pk)
+        try:
+            dashboard = get_object_or_404(queryset, pk=pk)
+        except Http404 as e:
+            user = cast(User, request.user)
+            alternative_team = Dashboard.objects.filter(deleted=False, team__in=user.teams)
```
Comment (paolodamico) [inline posthog/api/dashboard.py:196]: Aren't we missing the `pk` filter here?
```suggestion
            alternative_team = Dashboard.objects.filter(deleted=False, team__in=user.teams, pk=pk)
```

File: posthog/api/test/test_dashboard.py [hunk 2]
```diff
@@ -490,3 +490,22 @@ def test_insights_with_no_insight_set(self):
         )
         response = self.client.get(f"/api/projects/{self.team.id}/dashboards/{dashboard.pk}").json()
         self.assertEqual(response["items"][0]["filters"], {"events": [{"id": "$pageview"}], "insight": "TRENDS"})
+
+    def test_retrieve_dashboard_different_project_with_access(self):
```
Comment (paolodamico) [inline posthog/api/test/test_dashboard.py:494]: I would consider adding another test case where the dashboard ID exists but the user has no access to that team, to make sure the default 404 is sent (particularly if it comes from a private project in the same org too)

**Review by paolodamico — CHANGES_REQUESTED** [review]
Few comments inline before merging.

File: posthog/api/dashboard.py (hunk 1 above)
Comment (timgl) [reply posthog/api/dashboard.py:196]: 🤦

File: posthog/api/dashboard.py [hunk 3]
```diff
@@ -187,7 +189,18 @@ def get_queryset(self) -> QuerySet:
...
         queryset = self.get_queryset()
-        dashboard = get_object_or_404(queryset, pk=pk)
+        try:
+            dashboard = get_object_or_404(queryset, pk=pk)
+        except Http404 as e:
+            user = cast(User, request.user)
+            alternative_team = Dashboard.objects.filter(deleted=False, team__in=user.teams, pk=pk)
+            if alternative_team.exists():
+                raise ObjectExistsInOtherProject(
+                    alternative_team_id=alternative_team.first().team_id,  # type: ignore
```
Comment (Twixes) [inline posthog/api/dashboard.py:199]: This queries Postgres twice unnecessarily. The same could be achieved with just a `.first()`:
```suggestion
            alternative_team_dashboard = Dashboard.objects.filter(deleted=False, team__in=user.teams, pk=pk).first()
            if alternative_team_dashboard is not None:
                raise ObjectExistsInOtherProject(
                    alternative_team_id=alternative_team_dashboard.team_id,  # type: ignore
```

**Pull request merged**